compiler/rustc_typeck/src/astconv/generics.rs

```diff
@@ -57,19 +59,43 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             GenericArg::Const(_) => ParamKindOrd::Const { unordered },
         };
 
-        if matches!(arg, GenericArg::Type(hir::Ty { kind: hir::TyKind::Path { .. }, .. }))
-            && matches!(kind_ord, ParamKindOrd::Const { .. })
-        {
-            let suggestions = vec![
-                (arg.span().shrink_to_lo(), String::from("{ ")),
-                (arg.span().shrink_to_hi(), String::from(" }")),
-            ];
-            err.multipart_suggestion(
-                "if this generic argument was intended as a const parameter, \
-                try surrounding it with braces:",
-                suggestions,
-                Applicability::MaybeIncorrect,
-            );
+        // Specific suggestion set for diagnostics
+        if let Some(param) = param {
+            match (arg, &param.kind) {
+                (
+                    GenericArg::Type(hir::Ty { kind: hir::TyKind::Path { .. }, .. }),
+                    GenericParamDefKind::Const { .. },
+                ) => {
+                    let suggestions = vec![
+                        (arg.span().shrink_to_lo(), String::from("{ ")),
+                        (arg.span().shrink_to_hi(), String::from(" }")),
+                    ];
+                    err.multipart_suggestion(
+                        "if this generic argument was intended as a const parameter, \
+                    try surrounding it with braces:",
+                        suggestions,
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+                (
+                    GenericArg::Type(hir::Ty { kind: hir::TyKind::Array(_, len), .. }),
+                    GenericParamDefKind::Const { .. },
+                ) => {
+                    if tcx.type_of(param.def_id).is_ptr_sized_integral() {
+                        let snippet = sess
+                            .source_map()
+                            .span_to_snippet(tcx.hir().span(len.hir_id))
+                            .unwrap_or(String::from("N"));
+                        err.span_suggestion(
+                            arg.span(),
+                            "`[T; N]` provided where `usize` was expected, try",
```

Comment:
```suggestion
                            "`[T; N]` provided where `usize` was expected, try using the length directly",
```
might be a bit clearer

Reply:
I originally had `An array` was provided, but I thought explicitly describing the type was more clear